api/v1/verticadb_webhook.go

```diff
@@ -1701,6 +1728,194 @@ func (v *VerticaDB) checkSandboxPrimary(allErrs field.ErrorList, oldObj *Vertica
+	oldObj := old.(*VerticaDB)
+	persistSandboxes := v.findPersistSandboxes(oldObj)
+	oldSandboxMap := oldObj.GenSandboxMap()
+	newSandboxMap := v.GenSandboxMap()
+	newSandboxIndexMap := v.GenSandboxIndexMap()
+	newSubclusterMap := v.GenSubclusterMap()
+	statusSClusterIndexMap := v.GenStatusSClusterIndexMap()
+	for _, sandboxName := range persistSandboxes {
+		oldSandbox := oldSandboxMap[sandboxName]
+		newSandbox := newSandboxMap[sandboxName]
+		if oldSandbox.Image != newSandbox.Image {
+			newSandboxIndex := newSandboxIndexMap[sandboxName]
+			errMsg := v.checkSboxForShutdown(newSandbox, newSubclusterMap, statusSClusterIndexMap)
+			if errMsg != "" {
+				p := field.NewPath("spec").Child("sandboxes").Index(newSandboxIndex)
+				err := field.Invalid(p.Child("image"),
+					sandboxName,
+					fmt.Sprintf("cannot change the image for sandbox %q because %q",
+						sandboxName, errMsg))
+				allErrs = append(allErrs, err)
+			}
+		}
+	}
+	return allErrs
+}
+
+// ensure the sandbox to terminate is not shut down
+func (v *VerticaDB) validateTerminatingSandboxes(old runtime.Object, allErrs field.ErrorList) field.ErrorList {
+	oldObj := old.(*VerticaDB)
+	oldSandboxMap := oldObj.GenSandboxMap()
+	newSandboxMap := v.GenSandboxMap()
+	sandboxesToBeRemoved := vutil.MapKeyDiff(oldSandboxMap, newSandboxMap)
+	newSubclusterMap := v.GenSubclusterMap()
+	for _, sandboxToBeRemoved := range sandboxesToBeRemoved {
+		numOfSclustersInSbox := len(oldSandboxMap[sandboxToBeRemoved].Subclusters)
+		sclustersToTerminate := numOfSclustersInSbox
+		for _, subcluster := range oldSandboxMap[sandboxToBeRemoved].Subclusters {
+			if _, ok := newSubclusterMap[subcluster.Name]; ok {
+				sclustersToTerminate--
+			}
+		}
+		// sandboxToBeRemoved and all its subclusters are not found in new spec
+		if sclustersToTerminate == numOfSclustersInSbox {
+			if oldSandboxMap[sandboxToBeRemoved].Shutdown {
+				p := field.NewPath("spec").Child("sandboxes")
+				err := field.Invalid(p,
+					sandboxToBeRemoved,
+					fmt.Sprintf("cannot terminate sandbox %q that has Shutdown field set to true",
+						sandboxToBeRemoved))
+				allErrs = append(allErrs, err)
+			} // else looks good
+		} // else is expected
+	}
+	return allErrs
+}
+
+// checkSboxForShutdown checks if sbox or its scluster has shutdown field set to true in spec/status
+func (v *VerticaDB) checkSboxForShutdown(newSandbox *Sandbox, newSClusterMap map[string]*Subcluster, statusSClusterIndexMap map[string]int) string {
+	if newSandbox.Shutdown {
+		return fmt.Sprintf("Shutdown field of sandbox %q is set to true", newSandbox.Name)
+	}
+
+	for _, subcluster := range newSandbox.Subclusters {
+		i, foundInStatus := statusSClusterIndexMap[subcluster.Name]
+		if foundInStatus && (newSClusterMap[subcluster.Name].Shutdown || v.Status.Subclusters[i].Shutdown) {
+			return fmt.Sprintf("Subcluster %q is marked for shut down or has already been shut down", subcluster.Name)
+		}
+	}
+	return ""
```

Comment:
```suggestion
	return errMsgs
```

Reply:
same comment as the above one. When it returns empty string, it means no error.